docs/explanation/type-safety.md

```diff
@@ -0,0 +1,74 @@
+---
+title: Type safety
+---
+
+# Type safety
+
+React Router generates types for each route in your app that you can use to get type safety for each route module export.
+
+For example, let's say you have a `products/:id` route configured:
+
+```ts filename=app/routes.ts
+import {
+  type RouteConfig,
+  route,
+} from "@react-router/dev/routes";
+
+export const routes: RouteConfig = [
+  route("products/:id", "./routes/product.tsx"),
+];
+```
+
+Then, you can import route-specific types like so:
+
+```tsx filename=app/routes/product.tsx
+import type * as Route from "./+types.product";
+// types generated for this route 👆
+
+export function loader({ params }: Route.LoaderArgs) {
+  //                      👆 { id: string }
+  return { planet: `world #${params.id}` };
+}
+
+export default function Component({
+  loaderData, // 👈 { planet: string }
+}: Route.ComponentProps) {
+  return <h1>Hello, {loaderData.planet}!</h1>;
+}
+```
+
+If you haven't done so already, check out our guide for [setting up type safety][setting-up-type-safety] in a new project.
+
+## `typegen` command
+
+You can manually generate types with the `typegen` command:
+
+```sh
+react-router typegen
+```
+
+You can also use `--watch` to automatically regenerate types as files change:
+
+```sh
+react-router typegen --watch
+```
+
+The following types are generated for each route:
+
+- `LoaderArgs`
+- `ClientLoaderArgs`
+- `ActionArgs`
+- `ClientActionArgs`
+- `HydrateFallbackProps`
+- `ComponentProps` (for the `default` export)
+- `ErrorBoundaryProps`
+
+## How it works
+
+React Router's type generation executes your route config (`app/routes.ts` by default) to determine the routes for your app.
```

Comment:
What does "executes your route config" mean? Also, is `app/routes.ts` configurable? I thought it wasn't, and if not we shouldn't say "by default". If it is I just need to update my knowledge

Reply:
`app/` is configurable, and you can have other extensions for `route.ts` (`.tsx`, `.js`, `.jsx`) if you want

Reply:
"executes your route config" -> runs it with node (technically vite-node so that it can directly run `.ts` file) and looks at the `routes` export

Reply:
FWIW I've just been leaning into calling it `routes.ts` and `app/routes.ts` everywhere to avoid this sort of framing which can be confusing. Instead we could make sure to always link to a page with more detail about how it's more complicated than that?